db: TestExcise data race

```
==================
WARNING: DATA RACE
Write at 0x00c00577e68f by goroutine 44458:
  github.com/cockroachdb/pebble.TestExcise.func3()
      /home/runner/work/pebble/pebble/ingest_test.go:725 +0x826
  github.com/cockroachdb/datadriven.RunTest.func1()
      /home/runner/go/pkg/mod/github.com/cockroachdb/datadriven@v1.0.3-0.20240530155848-7682d40af056/datadriven.go:118 +0x58
  github.com/cockroachdb/datadriven.runDirective.func1()
      /home/runner/go/pkg/mod/github.com/cockroachdb/datadriven@v1.0.3-0.20240530155848-7682d40af056/datadriven.go:337 +0x10c
  github.com/cockroachdb/datadriven.runDirective()
      /home/runner/go/pkg/mod/github.com/cockroachdb/datadriven@v1.0.3-0.20240530155848-7682d40af056/datadriven.go:[34](https://github.com/cockroachdb/pebble/actions/runs/11073627431/job/30770572730?pr=3973#step:4:35)2 +0x8f
  github.com/cockroachdb/datadriven.runDirectiveOrSubTest()
      /home/runner/go/pkg/mod/github.com/cockroachdb/datadriven@v1.0.3-0.20240530155848-7682d40af056/datadriven.go:211 +0xa4
  github.com/cockroachdb/datadriven.runTestInternal()
      /home/runner/go/pkg/mod/github.com/cockroachdb/datadriven@v1.0.3-0.20240530155848-7682d40af056/datadriven.go:184 +0xe4
  github.com/cockroachdb/datadriven.RunTestAny()
      /home/runner/go/pkg/mod/github.com/cockroachdb/datadriven@v1.0.3-0.20240530155848-7682d40af056/datadriven.go:144 +0x3d0
  github.com/cockroachdb/datadriven.RunTest()
      /home/runner/go/pkg/mod/github.com/cockroachdb/datadriven@v1.0.3-0.20240530155848-7682d40af056/datadriven.go:117 +0xba
  github.com/cockroachdb/pebble.TestExcise()
      /home/runner/work/pebble/pebble/ingest_test.go:670 +0x5a4
  testing.tRunner()
      /opt/hostedtoolcache/go/1.22.7/x64/src/testing/testing.go:1689 +0x21e
  testing.(*T).Run.gowrap1()
      /opt/hostedtoolcache/go/1.22.7/x64/src/testing/testing.go:1742 +0x44

Previous write at 0x00c00577e68f by goroutine 44637:
  github.com/cockroachdb/pebble.TestExcise.func2.1()
      /home/runner/work/pebble/pebble/ingest_test.go:649 +0x2e
  github.com/cockroachdb/pebble.(*DB).flush1()
      /home/runner/work/pebble/pebble/compaction.go:1558 +0x1f34
  github.com/cockroachdb/pebble.(*DB).flush.func1()
      /home/runner/work/pebble/pebble/compaction.go:1163 +0x144
  runtime/pprof.Do()
      /opt/hostedtoolcache/go/1.22.7/x64/src/runtime/pprof/runtime.go:51 +0x117
  github.com/cockroachdb/pebble.(*DB).flush()
      /home/runner/work/pebble/pebble/compaction.go:1156 +0x91
  github.com/cockroachdb/pebble.(*DB).maybeScheduleFlush.gowrap1()
      /home/runner/work/pebble/pebble/compaction.go:1072 +0x33

Goroutine 44458 (running) created at:
  testing.(*T).Run()
      /opt/hostedtoolcache/go/1.22.7/x64/src/testing/testing.go:1742 +0x825
  testing.runTests.func1()
      /opt/hostedtoolcache/go/1.22.7/x64/src/testing/testing.go:2161 +0x85
  testing.tRunner()
      /opt/hostedtoolcache/go/1.22.7/x64/src/testing/testing.go:1689 +0x21e
  testing.runTests()
      /opt/hostedtoolcache/go/1.22.7/x64/src/testing/testing.go:2159 +0x8be
  testing.(*M).Run()
      /opt/hostedtoolcache/go/1.22.7/x64/src/testing/testing.go:2027 +0xf17
  main.main()
      _testmain.go:707 +0x2bd

Goroutine 44637 (running) created at:
  github.com/cockroachdb/pebble.(*DB).maybeScheduleFlush()
      /home/runner/work/pebble/pebble/compaction.go:1072 +0x186
  github.com/cockroachdb/pebble.(*DB).rotateMemtable()
      /home/runner/work/pebble/pebble/db.go:2621 +0x1f6
  github.com/cockroachdb/pebble.(*DB).handleIngestAsFlushable()
      /home/runner/work/pebble/pebble/ingest.go:1322 +0xce4
  github.com/cockroachdb/pebble.(*DB).ingest.func1()
      /home/runner/work/pebble/pebble/ingest.go:1577 +0x13dd
  github.com/cockroachdb/pebble.(*commitPipeline).AllocateSeqNum()
      /home/runner/work/pebble/pebble/commit.go:416 +0x709
  github.com/cockroachdb/pebble.(*DB).ingest()
      /home/runner/work/pebble/pebble/ingest.go:1639 +0x101d
  github.com/cockroachdb/pebble.(*DB).IngestAndExcise()
      /home/runner/work/pebble/pebble/ingest.go:1190 +0x45e
  github.com/cockroachdb/pebble.runIngestAndExciseCmd()
      /home/runner/work/pebble/pebble/data_test.go:1[35](https://github.com/cockroachdb/pebble/actions/runs/11073627431/job/30770572730?pr=3973#step:4:36)4 +0x699
  github.com/cockroachdb/pebble.TestExcise.func3()
      /home/runner/work/pebble/pebble/ingest_test.go:766 +0x165b
  github.com/cockroachdb/datadriven.RunTest.func1()
      /home/runner/go/pkg/mod/github.com/cockroachdb/datadriven@v1.0.3-0.20240530155848-7682d40af056/datadriven.go:118 +0x58
  github.com/cockroachdb/datadriven.runDirective.func1()
      /home/runner/go/pkg/mod/github.com/cockroachdb/datadriven@v1.0.3-0.20240530155848-7682d40af056/datadriven.go:3[37](https://github.com/cockroachdb/pebble/actions/runs/11073627431/job/30770572730?pr=3973#step:4:38) +0x10c
  github.com/cockroachdb/datadriven.runDirective()
      /home/runner/go/pkg/mod/github.com/cockroachdb/datadriven@v1.0.3-0.20240530155848-7682d40af056/datadriven.go:342 +0x8f
  github.com/cockroachdb/datadriven.runDirectiveOrSubTest()
      /home/runner/go/pkg/mod/github.com/cockroachdb/datadriven@v1.0.3-0.202[40](https://github.com/cockroachdb/pebble/actions/runs/11073627431/job/30770572730?pr=3973#step:4:41)530155848-7682d40af056/datadriven.go:211 +0xa4
  github.com/cockroachdb/datadriven.runTestInternal()
      /home/runner/go/pkg/mod/github.com/cockroachdb/datadriven@v1.0.3-0.20240530155848-7682d40af056/datadriven.go:184 +0xe4
  github.com/cockroachdb/datadriven.RunTestAny()
      /home/runner/go/pkg/mod/github.com/cockroachdb/datadriven@v1.0.3-0.20240530155848-7682d40af056/datadriven.go:144 +0x3d0
  github.com/cockroachdb/datadriven.RunTest()
      /home/runner/go/pkg/mod/github.com/cockroachdb/datadriven@v1.0.3-0.20240530155848-7682d40af056/datadriven.go:117 +0xba
  github.com/cockroachdb/pebble.TestExcise()
      /home/runner/work/pebble/pebble/ingest_test.go:670 +0x5a4
  testing.tRunner()
      /opt/hostedtoolcache/go/1.22.7/x64/src/testing/testing.go:1689 +0x21e
  testing.(*T).Run.gowrap1()
      /opt/hostedtoolcache/go/1.22.7/x64/src/testing/testing.go:17[42](https://github.com/cockroachdb/pebble/actions/runs/11073627431/job/30770572730?pr=3973#step:4:43) +0x44
==================
--- FAIL: TestExcise (0.20s)
```
https://github.com/cockroachdb/pebble/actions/runs/11073627431/job/30770572730?pr=3973

Jira issue: PEBBLE-266

--- a/ingest_test.go
+++ b/ingest_test.go
@@ -623,6 +623,17 @@ func TestExcise(t *testing.T) {
 		}
 		require.NoError(t, d.Close())
 	}()
+	clearFlushed := func() {
+		d.mu.Lock()
+		defer d.mu.Unlock()
+		// Wait for any ongoing flushes to stop first, otherwise the
+		// EventListener may set flushed=true due to a flush that was already in
+		// progress.
+		for d.mu.compact.flushing {
+			d.mu.compact.cond.Wait()
+		}
+		flushed = false
+	}
 
 	var opts *Options
 	reset := func(blockSize int) {
@@ -722,13 +733,9 @@ func TestExcise(t *testing.T) {
 			return ""
 
 		case "ingest":
-			flushed = false
-			noWait := false
-			for i := range td.CmdArgs {
-				switch td.CmdArgs[i].Key {
-				case "no-wait":
-					noWait = true
-				}
+			noWait := td.HasArg("no-wait")
+			if !noWait {
+				clearFlushed()
 			}
 			if err := runIngestCmd(td, d, mem); err != nil {
 				return err.Error()
@@ -748,16 +755,10 @@ func TestExcise(t *testing.T) {
 			return ""
 
 		case "ingest-and-excise":
-			flushed = false
-			noWait := false
-			for i := range td.CmdArgs {
-				switch td.CmdArgs[i].Key {
-				case "no-wait":
-					noWait = true
-				}
-			}
 			var prevFlushableIngests uint64
+			noWait := td.HasArg("no-wait")
 			if !noWait {
+				clearFlushed()
 				d.mu.Lock()
 				prevFlushableIngests = d.mu.versions.metrics.Flush.AsIngestCount
 				d.mu.Unlock()

